Gossiper: provide strong exception safety for endpoint state changes

bhalevy opened the pull request:
This series ensures that endpoint state changes (for each single endpoint) are applied to the gossiper endpoint_state_map as a whole and on all shards.
Any failure in the process will keep the existing endpoint state intact.

Note that verbs that modify the endpoint states of multiple endpoints may still succeed to modify some of them before hitting an error and those changes are committed to the endpoint_state_map, so we don't ensure atomicity when updating multiple endpoints' states.

Fixes scylladb/scylladb#14794
Fixes scylladb/scylladb#14799

bhalevy commented:
Converted to draft since this series depends on #14801

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3120/

bhalevy commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3120/

These are essentially https://github.com/scylladb/scylladb/issues/14799

bhalevy commented:
rebase (ready for review now)

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3283/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3285/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3317/

bhalevy commented:
In v2 ([66ec0ab](https://github.com/scylladb/scylladb/commit/66ec0ab43d6e93c82a527b94bbaa35bcc5c56eb0)):
- rebased
- fixed scylla-gdb to support endpoint_state_ptr in _endpoint_state_map

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3338/

bhalevy commented:
Marking as draft as the branch needs rebasing, and also it will depend on https://github.com/scylladb/scylladb/pull/15206

bhalevy commented:
In v3 ([5a8cb45](https://github.com/scylladb/scylladb/commit/5a8cb450cee3fbb35d85b168eb229546fc049995)):
- rebased (including the patches from https://github.com/scylladb/scylladb/pull/15206)
- dropped get_mutable_endpoint_state_ptr to help ensure endpoint state immutability
- added gossiper::update_timestamp as the standard way to update the endpoint_state update_timestamp in-place.
- dropped "gossiper: mark_dead: drop endpoint_state param"

tgrabiec commented on gms/gossiper.cc:
```diff
@@ -1327,14 +1327,16 @@ future<> gossiper::assassinate_endpoint(sstring address) {
-    return container().invoke_on(0, [endpoint] (auto&& gossiper) {
+future<generation_type> gossiper::get_current_generation_number(inet_address endpoint) const {
+    // FIXME: const container() has no const invoke_on variant
+    return const_cast<gossiper*>(this)->container().invoke_on(0, [endpoint] (const gossiper& gossiper) {
```
Why do you need get_current_generation_number() to be const?

bhalevy replied:
No need other than to accurately reflect what it does, or does not in this case.

bhalevy replied:
Will be deferred to a followup series

tgrabiec replied:
What I'm worried about is that marking things as const adds complications like in this FIXME, duplication of logic between const and non-const variants, and can put impediments to future extension when need to mutate something. So to pull its weight, it should be balanced by benefits in the form of more safety. In case of a gossiper, do we need a concept of constness? Maybe we don't.

bhalevy replied:
At an earlier version of this branch, this helped me sanitize the gossiper to make sure some const functions are guaranteed not to change _endpoint_state_map.

The const_cast here unfortunately was required due to limitation of our `container().invoke_on` templates that require a mutable instance for invoke_on.
I spent some time trying to come up with const overrides of these template, but I got tangled in the c++ templates web and decided to drop this effort :-/

bhalevy commented on gms/gossiper.cc:
```diff
@@ -699,6 +699,11 @@ future<> gossiper::remove_endpoint(inet_address endpoint, permit_id pid) {
+    std::optional<endpoint_state> state;
+    if (auto eps = get_endpoint_state_for_endpoint_ptr(endpoint)) {
+        state.emplace(*eps);
+    }
```
See #15206 (it is part of that series)
This was in response to this review comment: https://github.com/scylladb/scylladb/pull/15179#discussion_r1306969125

The optional<endpoint_state> is for safety, I agree the if the node was alive it should always have an endpoint_state (see also the extra verification in `gossiper::is_alive`)

bhalevy commented on gms/gossiper.cc:
```diff
@@ -2328,12 +2328,12 @@ bool gossiper::is_alive(inet_address ep) const {
 }
 
-future<> gossiper::wait_alive(std::vector<gms::inet_address> nodes, std::chrono::milliseconds timeout) {
+future<> gossiper::wait_alive(std::vector<gms::inet_address> nodes, std::chrono::milliseconds timeout) const {
```
This one is just for cleanup. Since both functions do not modify the gossiper state they can be marked as const.
In general, I'd ask why NOT mark it as const.
There needs to be a reason to change the object to do so.
it shouldn't be the default as it makes it harder to implement methods that need to be const as they can't just call other mutable methods.

bhalevy replied:
I'll defer the cleanups to a followup PR

bhalevy commented on gms/gossiper.cc:
```diff
@@ -582,23 +582,24 @@ future<> gossiper::do_apply_state_locally(gms::inet_address node, const endpoint
                     logger.debug("Ignoring remote version {} <= {} for {}", remote_max_version, local_max_version, node);
                 }
-                if (!is_alive(node) && !this->is_dead_state(local_state)) { // unless of course, it was dead
+                if (should_mark_alive) { // unless of course, it was dead
```
Good point.

We can fetch the endpoint_state again after apply_new_states, which isn't too bad,
or the call to `is_dead_state` that was moved here before `apply_new_states` can check `remote_state`, not `local_state`.

We can also go out of our way and let replicate and then apply_new_states and the like return an endpoint_state_ptr of the modified state, that we can use here, after it was updated. But that may be an overkill.

bhalevy commented:
In v4 ([98fd9fc](https://github.com/scylladb/scylladb/commit/98fd9fcc11f264c8f6f644c15fa31c4a9134abaf)):
- rebased
- dropped non-functional cleanup patches (will be submitted separately)
- gossiper: move endpoint_state by value to apply it
  - lookup the endpoint_state after apply the remote changes onto it for `!this->is_dead_state`, as per https://github.com/scylladb/scylladb/pull/15073#discussion_r1310821564

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3383/
